upb/mini_table.c

```diff
@@ -0,0 +1,937 @@
+    char ch = *ptr++;
+    if (ch <= kUpb_EncodedValue_MaxField) {
+      upb_MiniTable_Field* field = fields;
+      *field_count += 1;
+      fields = (char*)fields + field_size;
+      field->number = ++last_field_number;
+      upb_MiniTable_SetField(d, ch, field, msg_modifiers, sub_count);
+    } else if (kUpb_EncodedValue_MinModifier <= ch &&
+               ch <= kUpb_EncodedValue_MaxModifier) {
+      ptr = upb_MtDecoder_ParseModifier(d, ptr, ch, *field_count, &msg_modifiers);
+    } else if (ch == kUpb_EncodedValue_End) {
+      if (!d->table) {
+        upb_MtDecoder_ErrorFormat(d, "Extensions cannot have oneofs.");
+        UPB_UNREACHABLE();
+      }
+      ptr = upb_MtDecoder_DecodeOneofs(d, ptr);
+    } else if (kUpb_EncodedValue_MinSkip <= ch &&
+               ch <= kUpb_EncodedValue_MaxSkip) {
+      if (need_dense_below) {
+        d->table->dense_below = d->table->field_count;
+        need_dense_below = false;
+      }
+      uint32_t skip;
+      ptr = upb_MiniTable_DecodeBase92Varint(d, ptr, ch,
+                                             kUpb_EncodedValue_MinSkip,
+                                             kUpb_EncodedValue_MaxSkip, &skip);
+      last_field_number += skip;
+      last_field_number--;  // Next field seen will increment.
+    }
+  }
+
+  if (need_dense_below) {
+    d->table->dense_below = d->table->field_count;
+  }
+}
+
+static void upb_MtDecoder_ParseMessage(upb_MtDecoder* d, const char* data,
+                                       size_t len) {
+  // Buffer length is an upper bound on the number of fields. We will return
+  // what we don't use.
+  d->fields = upb_Arena_Malloc(d->arena, sizeof(*d->fields) * len);
+  upb_MtDecoder_CheckOutOfMemory(d, d->fields);
+
+  uint32_t sub_count = 0;
+  d->table->field_count = 0;
+  d->table->fields = d->fields;
+  upb_MtDecoder_Parse(d, data, len, d->fields, sizeof(*d->fields),
+                      &d->table->field_count, &sub_count);
+
+  // Return unused memory from fields array.
+  assert(d->table->field_count <= len);
+  d->fields = upb_Arena_Realloc(d->arena, d->fields, sizeof(*d->fields) * len,
+                                sizeof(*d->fields) * d->table->field_count);
+  d->table->fields = d->fields;
+  upb_MtDecoder_AllocateSubs(d, sub_count);
+}
+
+#define UPB_COMPARE_INTEGERS(a, b) ((a) < (b) ? -1 : ((a) == (b) ? 0 : 1))
+
+int upb_MtDecoder_CompareFields(const void* _a, const void* _b) {
+  const upb_LayoutItem* a = _a;
+  const upb_LayoutItem* b = _b;
+  // Currently we just sort by:
+  //  1. rep (descending, so largest fields are first)
+  //  2. is_case (descending, so oneof cases are first)
+  //  2. field_number (ascending, so smallest numbers are first)
+  //
+  // The main goal of this is to reduce space lost to padding.
+  if (a->rep != b->rep) return UPB_COMPARE_INTEGERS(a->rep, b->rep);
+  if (a->type != b->type) return UPB_COMPARE_INTEGERS(a->type, b->type);
+  return UPB_COMPARE_INTEGERS(b->field_index, a->field_index);
```

Comment:
if you are feeling clever, I think you can use math to avoid some if statements.  Because you have bounds on reps, I think you can do:

uint64 a = a->ref * REP_MAX + a->type * TYPE_MAX + a->field_index;

and only compare the result

Reply:
Done.